File: projects/simple-compiler/src/lexer.rs
/// Token types for the Simple Language
#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    // Literals
    Integer(i64),
    Float(f64),
    StringLiteral(String),
    Identifier(String),

    // Keywords
    Let,
    If,
    Else,
    While,
    Fn,
    Return,
    Print,
    True,
    False,

    // Operators
    Plus,         // +
    Minus,        // -
    Star,         // *
    Slash,        // /
    Percent,      // %
    Equal,        // =
    EqualEqual,   // ==
    BangEqual,    // !=
    Less,         // <
    LessEqual,    // <=
    Greater,      // >
    GreaterEqual, // >=
    Bang,         // !
    AmpAmp,       // &&
    PipePipe,     // ||

    // Delimiters
    LeftParen,    // (
    RightParen,   // )
    LeftBrace,    // {
    RightBrace,   // }
    Comma,        // ,
    Semicolon,    // ;
    Colon,        // :
    Arrow,        // ->

    // Special
    Eof,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(kind: TokenKind, line: usize, column: usize) -> Self {
        Self { kind, line, column }
    }
}

/// Lexer - converts source code into a stream of tokens
///
/// The lexer is the first phase of compilation. It reads the source code
/// character by character and groups them into meaningful tokens.
pub struct Lexer {
    source: Vec<char>,
    pos: usize,
    line: usize,
    column: usize,
}

impl Lexer {
    pub fn new(source: &str) -> Self {
        Self {
            source: source.chars().collect(),
            pos: 0,
            line: 1,
            column: 1,
        }
    }

    /// Tokenize the entire source code
    pub fn tokenize(&mut self) -> Result<Vec<Token>, String> {
        let mut tokens = Vec::new();

        loop {
            self.skip_whitespace_and_comments();
            if self.is_at_end() {
                tokens.push(Token::new(TokenKind::Eof, self.line, self.column));
                break;
            }

            let token = self.next_token()?;
            tokens.push(token);
        }

        Ok(tokens)
    }

    fn next_token(&mut self) -> Result<Token, String> {
        let line = self.line;
        let column = self.column;
        let ch = self.advance();

        match ch {
            // Single-character tokens
            '(' => Ok(Token::new(TokenKind::LeftParen, line, column)),
            ')' => Ok(Token::new(TokenKind::RightParen, line, column)),
            '{' => Ok(Token::new(TokenKind::LeftBrace, line, column)),
            '}' => Ok(Token::new(TokenKind::RightBrace, line, column)),
            ',' => Ok(Token::new(TokenKind::Comma, line, column)),
            ';' => Ok(Token::new(TokenKind::Semicolon, line, column)),
            ':' => Ok(Token::new(TokenKind::Colon, line, column)),
            '+' => Ok(Token::new(TokenKind::Plus, line, column)),
            '*' => Ok(Token::new(TokenKind::Star, line, column)),
            '%' => Ok(Token::new(TokenKind::Percent, line, column)),

            // Two-character tokens
            '-' => {
                if self.match_char('>') {
                    Ok(Token::new(TokenKind::Arrow, line, column))
                } else {
                    Ok(Token::new(TokenKind::Minus, line, column))
                }
            }
            '=' => {
                if self.match_char('=') {
                    Ok(Token::new(TokenKind::EqualEqual, line, column))
                } else {
                    Ok(Token::new(TokenKind::Equal, line, column))
                }
            }
            '!' => {
                if self.match_char('=') {
                    Ok(Token::new(TokenKind::BangEqual, line, column))
                } else {
                    Ok(Token::new(TokenKind::Bang, line, column))
                }
            }
            '<' => {
                if self.match_char('=') {
                    Ok(Token::new(TokenKind::LessEqual, line, column))
                } else {
                    Ok(Token::new(TokenKind::Less, line, column))
                }
            }
            '>' => {
                if self.match_char('=') {
                    Ok(Token::new(TokenKind::GreaterEqual, line, column))
                } else {
                    Ok(Token::new(TokenKind::Greater, line, column))
                }
            }
            '&' => {
                if self.match_char('&') {
                    Ok(Token::new(TokenKind::AmpAmp, line, column))
                } else {
                    Err(format!("Unexpected character '&' at {}:{}", line, column))
                }
            }
            '|' => {
                if self.match_char('|') {
                    Ok(Token::new(TokenKind::PipePipe, line, column))
                } else {
                    Err(format!("Unexpected character '|' at {}:{}", line, column))
                }
            }
            '/' => Ok(Token::new(TokenKind::Slash, line, column)),

            // String literals
            '"' => self.read_string(line, column),

            // Numbers and identifiers
            _ if ch.is_ascii_digit() => self.read_number(ch, line, column),
            _ if ch.is_ascii_alphabetic() || ch == '_' => self.read_identifier(ch, line, column),

            _ => Err(format!("Unexpected character '{}' at {}:{}", ch, line, column)),
        }
    }

    fn read_string(&mut self, line: usize, column: usize) -> Result<Token, String> {
        let mut value = String::new();
        loop {
            if self.is_at_end() {
                return Err(format!("Unterminated string at {}:{}", line, column));
            }
            let ch = self.advance();
            if ch == '"' {
                break;
            }
            if ch == '\\' {
                let escaped = self.advance();
                match escaped {
                    'n' => value.push('\n'),
                    't' => value.push('\t'),
                    '\\' => value.push('\\'),
                    '"' => value.push('"'),
                    _ => return Err(format!("Invalid escape sequence '\\{}' at {}:{}", escaped, line, column)),
                }
            } else {
                value.push(ch);
            }
        }
        Ok(Token::new(TokenKind::StringLiteral(value), line, column))
    }

    fn read_number(&mut self, first: char, line: usize, column: usize) -> Result<Token, String> {
        let mut text = String::from(first);
        let mut is_float = false;

        while !self.is_at_end() && (self.peek().is_ascii_digit() || self.peek() == '.') {
            if self.peek() == '.' {
                if is_float {
                    return Err(format!("Invalid number at {}:{}", line, column));
                }
                is_float = true;
            }
            text.push(self.advance());
        }

        if is_float {
            text.parse::<f64>()
                .map(|v| Token::new(TokenKind::Float(v), line, column))
                .map_err(|_| format!("Invalid float at {}:{}", line, column))
        } else {
            text.parse::<i64>()
                .map(|v| Token::new(TokenKind::Integer(v), line, column))
                .map_err(|_| format!("Invalid integer at {}:{}", line, column))
        }
    }

    fn read_identifier(&mut self, first: char, line: usize, column: usize) -> Result<Token, String> {
        let mut text = String::from(first);
        while !self.is_at_end() && (self.peek().is_ascii_alphanumeric() || self.peek() == '_') {
            text.push(self.advance());
        }

        let kind = match text.as_str() {
            "let" => TokenKind::Let,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "while" => TokenKind::While,
            "fn" => TokenKind::Fn,
            "return" => TokenKind::Return,
            "print" => TokenKind::Print,
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            _ => TokenKind::Identifier(text),
        };

        Ok(Token::new(kind, line, column))
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            if self.is_at_end() {
                return;
            }
            match self.peek() {
                ' ' | '\t' | '\r' => {
                    self.advance();
                }
                '\n' => {
                    self.advance();
                    self.line += 1;
                    self.column = 1;
                }
                '/' if self.peek_next() == '/' => {
                    // Line comment
                    while !self.is_at_end() && self.peek() != '\n' {
                        self.advance();
                    }
                }
                '/' if self.peek_next() == '*' => {
                    // Block comment
                    self.advance(); // skip /
                    self.advance(); // skip *
                    let mut depth = 1;
                    while depth > 0 && !self.is_at_end() {
                        if self.peek() == '/' && self.peek_next() == '*' {
                            self.advance();
                            self.advance();
                            depth += 1;
                        } else if self.peek() == '*' && self.peek_next() == '/' {
                            self.advance();
                            self.advance();
                            depth -= 1;
                        } else {
                            if self.peek() == '\n' {
                                self.line += 1;
                                self.column = 0;
                            }
                            self.advance();
                        }
                    }
                }
                _ => return,
            }
        }
    }

    fn advance(&mut self) -> char {
        let ch = self.source[self.pos];
        self.pos += 1;
        self.column += 1;
        ch
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            '\0'
        } else {
            self.source[self.pos]
        }
    }

    fn peek_next(&self) -> char {
        if self.pos + 1 >= self.source.len() {
            '\0'
        } else {
            self.source[self.pos + 1]
        }
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() || self.source[self.pos] != expected {
            false
        } else {
            self.advance();
            true
        }
    }

    fn is_at_end(&self) -> bool {
        self.pos >= self.source.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_source() {
        let mut lexer = Lexer::new("");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].kind, TokenKind::Eof);
    }

    #[test]
    fn test_single_tokens() {
        let mut lexer = Lexer::new("+ - * / ( ) { } ; =");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Plus);
        assert_eq!(tokens[1].kind, TokenKind::Minus);
        assert_eq!(tokens[2].kind, TokenKind::Star);
        assert_eq!(tokens[3].kind, TokenKind::Slash);
        assert_eq!(tokens[4].kind, TokenKind::LeftParen);
        assert_eq!(tokens[5].kind, TokenKind::RightParen);
        assert_eq!(tokens[6].kind, TokenKind::LeftBrace);
        assert_eq!(tokens[7].kind, TokenKind::RightBrace);
        assert_eq!(tokens[8].kind, TokenKind::Semicolon);
        assert_eq!(tokens[9].kind, TokenKind::Equal);
    }

    #[test]
    fn test_two_char_operators() {
        let mut lexer = Lexer::new("== != <= >= && || ->");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::EqualEqual);
        assert_eq!(tokens[1].kind, TokenKind::BangEqual);
        assert_eq!(tokens[2].kind, TokenKind::LessEqual);
        assert_eq!(tokens[3].kind, TokenKind::GreaterEqual);
        assert_eq!(tokens[4].kind, TokenKind::AmpAmp);
        assert_eq!(tokens[5].kind, TokenKind::PipePipe);
        assert_eq!(tokens[6].kind, TokenKind::Arrow);
    }

    #[test]
    fn test_integer_literal() {
        let mut lexer = Lexer::new("42 0 100000");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Integer(42));
        assert_eq!(tokens[1].kind, TokenKind::Integer(0));
        assert_eq!(tokens[2].kind, TokenKind::Integer(100000));
    }

    #[test]
    fn test_float_literal() {
        let mut lexer = Lexer::new("3.14 0.5 100.0");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Float(3.14));
        assert_eq!(tokens[1].kind, TokenKind::Float(0.5));
        assert_eq!(tokens[2].kind, TokenKind::Float(100.0));
    }

    #[test]
    fn test_string_literal() {
        let mut lexer = Lexer::new(r#""hello" "world\n""#);
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::StringLiteral("hello".to_string()));
        assert_eq!(tokens[1].kind, TokenKind::StringLiteral("world\n".to_string()));
    }

    #[test]
    fn test_keywords() {
        let mut lexer = Lexer::new("let if else while fn return print true false");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Let);
        assert_eq!(tokens[1].kind, TokenKind::If);
        assert_eq!(tokens[2].kind, TokenKind::Else);
        assert_eq!(tokens[3].kind, TokenKind::While);
        assert_eq!(tokens[4].kind, TokenKind::Fn);
        assert_eq!(tokens[5].kind, TokenKind::Return);
        assert_eq!(tokens[6].kind, TokenKind::Print);
        assert_eq!(tokens[7].kind, TokenKind::True);
        assert_eq!(tokens[8].kind, TokenKind::False);
    }

    #[test]
    fn test_identifiers() {
        let mut lexer = Lexer::new("foo bar_baz x1");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Identifier("foo".to_string()));
        assert_eq!(tokens[1].kind, TokenKind::Identifier("bar_baz".to_string()));
        assert_eq!(tokens[2].kind, TokenKind::Identifier("x1".to_string()));
    }

    #[test]
    fn test_comments() {
        let mut lexer = Lexer::new("42 // this is a comment\n10");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Integer(42));
        assert_eq!(tokens[1].kind, TokenKind::Integer(10));
    }

    #[test]
    fn test_block_comment() {
        let mut lexer = Lexer::new("42 /* block\ncomment */ 10");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Integer(42));
        assert_eq!(tokens[1].kind, TokenKind::Integer(10));
    }

    #[test]
    fn test_let_statement() {
        let mut lexer = Lexer::new("let x = 42;");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].kind, TokenKind::Let);
        assert_eq!(tokens[1].kind, TokenKind::Identifier("x".to_string()));
        assert_eq!(tokens[2].kind, TokenKind::Equal);
        assert_eq!(tokens[3].kind, TokenKind::Integer(42));
        assert_eq!(tokens[4].kind, TokenKind::Semicolon);
    }

    #[test]
    fn test_error_unterminated_string() {
        let mut lexer = Lexer::new(r#""hello"#);
        assert!(lexer.tokenize().is_err());
    }

    #[test]
    fn test_error_invalid_character() {
        let mut lexer = Lexer::new("@");
        assert!(lexer.tokenize().is_err());
    }

    #[test]
    fn test_line_and_column_tracking() {
        let mut lexer = Lexer::new("let\n  x = 42;");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].line, 1);
        assert_eq!(tokens[0].column, 1);
        assert_eq!(tokens[1].line, 2);
        assert_eq!(tokens[1].column, 3);
    }
}

File: projects/simple-compiler/src/lexer.cpp
#include "lexer.hpp"
#include <cctype>
#include <stdexcept>

namespace compiler {

Lexer::Lexer(const std::string& source)
    : source_(source), current_(0), line_(1), column_(1) {}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;

    while (!isAtEnd()) {
        skipWhitespace();

        if (isAtEnd()) break;

        // 跳过注释
        if (skipComment()) {
            continue;
        }

        size_t start = current_;
        int startLine = line_;
        int startColumn = column_;

        char c = advance();

        // 标识符或关键字
        if (isAlpha(c)) {
            // 回退一个字符，让identifier()处理
            current_--;
            column_--;
            tokens.push_back(identifier());
        }
        // 数字
        else if (isDigit(c)) {
            current_--;
            column_--;
            tokens.push_back(number());
        }
        // 字符串
        else if (c == '"') {
            current_--;
            column_--;
            tokens.push_back(string());
        }
        // 单引号字符串
        else if (c == '\'') {
            current_--;
            column_--;
            tokens.push_back(string());
        }
        // 运算符和分隔符
        else {
            current_--;
            column_--;
            Token token = operatorToken();
            if (token.type != TokenType::NEWLINE) {
                tokens.push_back(token);
            }
        }
    }

    // 添加EOF token
    tokens.emplace_back(TokenType::EOF_TOKEN, "", line_, column_);
    return tokens;
}

Token Lexer::nextToken() {
    skipWhitespace();

    if (isAtEnd()) {
        return Token(TokenType::EOF_TOKEN, "", line_, column_);
    }

    // 跳过注释
    while (skipComment()) {
        skipWhitespace();
    }

    if (isAtEnd()) {
        return Token(TokenType::EOF_TOKEN, "", line_, column_);
    }

    char c = advance();

    // 标识符或关键字
    if (isAlpha(c)) {
        current_--;
        column_--;
        return identifier();
    }
    // 数字
    else if (isDigit(c)) {
        current_--;
        column_--;
        return number();
    }
    // 字符串
    else if (c == '"' || c == '\'') {
        current_--;
        column_--;
        return string();
    }
    // 运算符和分隔符
    else {
        current_--;
        column_--;
        return operatorToken();
    }
}

bool Lexer::hasMore() const {
    return !isAtEnd();
}

char Lexer::peek() const {
    if (isAtEnd()) return '\0';
    return source_[current_];
}

char Lexer::peekNext() const {
    if (current_ + 1 >= source_.size()) return '\0';
    return source_[current_ + 1];
}

char Lexer::advance() {
    char c = source_[current_++];
    if (c == '\n') {
        line_++;
        column_ = 1;
    } else {
        column_++;
    }
    return c;
}

bool Lexer::match(char expected) {
    if (isAtEnd()) return false;
    if (source_[current_] != expected) return false;

    current_++;
    if (expected == '\n') {
        line_++;
        column_ = 1;
    } else {
        column_++;
    }
    return true;
}

void Lexer::skipWhitespace() {
    while (!isAtEnd()) {
        char c = peek();
        if (c == ' ' || c == '\t' || c == '\r') {
            advance();
        } else if (c == '\n') {
            // 可选：将换行作为token
            advance();
        } else {
            break;
        }
    }
}

bool Lexer::skipComment() {
    if (isAtEnd()) return false;

    // 单行注释 //
    if (peek() == '/' && peekNext() == '/') {
        advance(); // 消耗第一个 /
        advance(); // 消耗第二个 /
        while (!isAtEnd() && peek() != '\n') {
            advance();
        }
        return true;
    }

    // 多行注释 /* ... */
    if (peek() == '/' && peekNext() == '*') {
        int startLine = line_;
        advance(); // 消耗 /
        advance(); // 消耗 *
        while (!isAtEnd()) {
            if (peek() == '*' && peekNext() == '/') {
                advance(); // 消耗 *
                advance(); // 消耗 /
                return true;
            }
            advance();
        }
        errors_.push_back("Unterminated comment starting at line " +
                          std::to_string(startLine));
        return true;
    }

    return false;
}

bool Lexer::isAtEnd() const {
    return current_ >= source_.size();
}

Token Lexer::makeToken(TokenType type, const std::string& lexeme) {
    return Token(type, lexeme, line_, column_ - lexeme.length());
}

Token Lexer::makeToken(TokenType type, const std::string& lexeme, TokenValue value) {
    return Token(type, lexeme, value, line_, column_ - lexeme.length());
}

Token Lexer::errorToken(const std::string& message) {
    errors_.push_back(message + " at line " + std::to_string(line_) +
                      ", column " + std::to_string(column_));
    return Token(TokenType::ERROR, message, line_, column_);
}

Token Lexer::identifier() {
    size_t start = current_;
    int startColumn = column_;

    while (!isAtEnd() && isAlphaNumeric(peek())) {
        advance();
    }

    std::string text = source_.substr(start, current_ - start);

    // 检查是否是关键字
    const auto& keywords = getKeywords();
    auto it = keywords.find(text);
    if (it != keywords.end()) {
        return Token(it->second, text, line_, startColumn);
    }

    return Token(TokenType::IDENTIFIER, text, line_, startColumn);
}

Token Lexer::number() {
    size_t start = current_;
    int startColumn = column_;
    bool isFloat = false;

    // 整数部分
    while (!isAtEnd() && isDigit(peek())) {
        advance();
    }

    // 小数部分
    if (!isAtEnd() && peek() == '.' && isDigit(peekNext())) {
        isFloat = true;
        advance(); // 消耗 .
        while (!isAtEnd() && isDigit(peek())) {
            advance();
        }
    }

    // 科学计数法
    if (!isAtEnd() && (peek() == 'e' || peek() == 'E')) {
        isFloat = true;
        advance();
        if (!isAtEnd() && (peek() == '+' || peek() == '-')) {
            advance();
        }
        while (!isAtEnd() && isDigit(peek())) {
            advance();
        }
    }

    std::string text = source_.substr(start, current_ - start);

    if (isFloat) {
        try {
            double value = std::stod(text);
            return Token(TokenType::FLOAT, text, value, line_, startColumn);
        } catch (const std::exception& e) {
            return errorToken("Invalid float literal: " + text);
        }
    } else {
        try {
            int64_t value = std::stoll(text);
            return Token(TokenType::INTEGER, text, value, line_, startColumn);
        } catch (const std::exception& e) {
            return errorToken("Invalid integer literal: " + text);
        }
    }
}

Token Lexer::string() {
    int startLine = line_;
    int startColumn = column_;
    char quote = advance(); // 消耗引号
    std::string value;

    while (!isAtEnd() && peek() != quote) {
        if (peek() == '\\') {
            advance(); // 消耗反斜杠
            if (isAtEnd()) break;

            char escaped = advance();
            switch (escaped) {
                case 'n': value += '\n'; break;
                case 't': value += '\t'; break;
                case 'r': value += '\r'; break;
                case '\\': value += '\\'; break;
                case '\'': value += '\''; break;
                case '"': value += '"'; break;
                case '0': value += '\0'; break;
                default:
                    value += '\\';
                    value += escaped;
                    break;
            }
        } else {
            value += advance();
        }
    }

    if (isAtEnd()) {
        return errorToken("Unterminated string starting at line " +
                          std::to_string(startLine));
    }

    advance(); // 消耗结束引号
    std::string lexeme = source_.substr(startColumn - 1, current_ - startColumn + 1);
    return Token(TokenType::STRING, lexeme, value, startLine, startColumn);
}

Token Lexer::operatorToken() {
    int startLine = line_;
    int startColumn = column_;
    char c = advance();

    switch (c) {
        case '+':
            if (match('+')) return Token(TokenType::INCREMENT, "++", startLine, startColumn);
            if (match('=')) return Token(TokenType::PLUS_ASSIGN, "+=", startLine, startColumn);
            return Token(TokenType::PLUS, "+", startLine, startColumn);

        case '-':
            if (match('-')) return Token(TokenType::DECREMENT, "--", startLine, startColumn);
            if (match('=')) return Token(TokenType::MINUS_ASSIGN, "-=", startLine, startColumn);
            if (match('>')) return Token(TokenType::ARROW, "->", startLine, startColumn);
            return Token(TokenType::MINUS, "-", startLine, startColumn);

        case '*':
            if (match('*')) return Token(TokenType::POWER, "**", startLine, startColumn);
            if (match('=')) return Token(TokenType::MUL_ASSIGN, "*=", startLine, startColumn);
            return Token(TokenType::MULTIPLY, "*", startLine, startColumn);

        case '/':
            if (match('=')) return Token(TokenType::DIV_ASSIGN, "/=", startLine, startColumn);
            return Token(TokenType::DIVIDE, "/", startLine, startColumn);

        case '%':
            return Token(TokenType::MODULO, "%", startLine, startColumn);

        case '=':
            if (match('=')) return Token(TokenType::EQUAL, "==", startLine, startColumn);
            return Token(TokenType::ASSIGN, "=", startLine, startColumn);

        case '!':
            if (match('=')) return Token(TokenType::NOT_EQUAL, "!=", startLine, startColumn);
            return Token(TokenType::NOT, "!", startLine, startColumn);

        case '<':
            if (match('=')) return Token(TokenType::LESS_EQUAL, "<=", startLine, startColumn);
            if (match('<')) return Token(TokenType::SHIFT_LEFT, "<<", startLine, startColumn);
            return Token(TokenType::LESS, "<", startLine, startColumn);

        case '>':
            if (match('=')) return Token(TokenType::GREATER_EQUAL, ">=", startLine, startColumn);
            if (match('>')) return Token(TokenType::SHIFT_RIGHT, ">>", startLine, startColumn);
            return Token(TokenType::GREATER, ">", startLine, startColumn);

        case '&':
            if (match('&')) return Token(TokenType::AND, "&&", startLine, startColumn);
            return Token(TokenType::BIT_AND, "&", startLine, startColumn);

        case '|':
            if (match('|')) return Token(TokenType::OR, "||", startLine, startColumn);
            return Token(TokenType::BIT_OR, "|", startLine, startColumn);

        case '^':
            return Token(TokenType::BIT_XOR, "^", startLine, startColumn);

        case '~':
            return Token(TokenType::BIT_NOT, "~", startLine, startColumn);

        case '(':
            return Token(TokenType::LEFT_PAREN, "(", startLine, startColumn);

        case ')':
            return Token(TokenType::RIGHT_PAREN, ")", startLine, startColumn);

        case '{':
            return Token(TokenType::LEFT_BRACE, "{", startLine, startColumn);

        case '}':
            return Token(TokenType::RIGHT_BRACE, "}", startLine, startColumn);

        case '[':
            return Token(TokenType::LEFT_BRACKET, "[", startLine, startColumn);

        case ']':
            return Token(TokenType::RIGHT_BRACKET, "]", startLine, startColumn);

        case ',':
            return Token(TokenType::COMMA, ",", startLine, startColumn);

        case '.':
            return Token(TokenType::DOT, ".", startLine, startColumn);

        case ';':
            return Token(TokenType::SEMICOLON, ";", startLine, startColumn);

        case ':':
            if (match(':')) return Token(TokenType::DOUBLE_COLON, "::", startLine, startColumn);
            return Token(TokenType::COLON, ":", startLine, startColumn);

        case '\n':
            return Token(TokenType::NEWLINE, "\\n", startLine, startColumn);

        default:
            return errorToken(std::string("Unexpected character: ") + c);
    }
}

bool Lexer::isAlpha(char c) const {
    return std::isalpha(c) || c == '_';
}

bool Lexer::isDigit(char c) const {
    return std::isdigit(c);
}

bool Lexer::isAlphaNumeric(char c) const {
    return isAlpha(c) || isDigit(c);
}

} // namespace compiler
